File: packages/telemetry/src/__tests__/promql-builder.test.ts
import { describe, expect, it } from 'vitest';
import {
	buildErrorRateQuery,
	buildLatencyP95Query,
	buildMetricQuery,
	buildSampleCountQuery,
	escapePromqlLabelValue,
} from '../query/build-promql.js';
import { UnsupportedMetricTypeError } from '../errors/telemetry-adapter-error.js';

const policy = {
	metricType: 'error_rate',
	threshold: 0.01,
	serviceName: 'demo-service',
	windowSeconds: 300,
};

const runContext = {
	flagKey: 'demo-feature-flag',
	treatmentVariationId: 'treatment-var',
	controlVariationId: 'control-var',
};

describe('build-promql', () => {
	it('escapes quotes and backslashes in label values', () => {
		expect(escapePromqlLabelValue('a"b\\c')).toBe('a\\"b\\\\c');
	});

	it('builds error rate query with user context and window', () => {
		const query = buildErrorRateQuery(policy, runContext, 'treatment');
		expect(query).toContain('ld_context_kind="user"');
		expect(query).toContain('ld_variation_id="treatment-var"');
		expect(query).toContain('[300s]');
		expect(query).toContain('status=~"5.."');
	});

	it('builds latency p95 query in milliseconds', () => {
		const latencyPolicy = { ...policy, metricType: 'latency_p95' };
		const query = buildLatencyP95Query(latencyPolicy, runContext, 'control');
		expect(query).toContain('histogram_quantile(0.95');
		expect(query).toContain('ld_variation_id="control-var"');
		expect(query).toContain('* 1000');
	});

	it('builds sample count query for cohorts', () => {
		const query = buildSampleCountQuery(policy, runContext, 'treatment');
		expect(query).toContain('sum(increase(http_requests_total');
		expect(query).toContain('ld_variation_id="treatment-var"');
	});

	it('supports canonical metric types only', () => {
		expect(buildMetricQuery('error_rate', policy, runContext, 'treatment')).toContain(
			'rate(http_requests_total',
		);
		expect(
			buildMetricQuery('latency_p95', { ...policy, metricType: 'latency_p95' }, runContext, 'treatment'),
		).toContain('histogram_quantile');
		expect(() =>
			buildMetricQuery('p95_latency_ms', policy, runContext, 'treatment'),
		).toThrow(UnsupportedMetricTypeError);
	});
});

File: kotlin/modules/telemetry/src/test/kotlin/com/ffpromo/telemetry/query/PromqlBuilderTest.kt
package com.ffpromo.telemetry.query

import com.ffpromo.contracts.GatePolicyInput
import com.ffpromo.contracts.GateRunContext
import com.ffpromo.contracts.MetricType
import com.ffpromo.telemetry.errors.UnsupportedMetricTypeError
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class PromqlBuilderTest {
    private val policy = GatePolicyInput(
        metricType = MetricType.error_rate,
        threshold = 0.01,
        serviceName = "demo-service",
        windowSeconds = 300,
    )

    private val runContext = GateRunContext(
        flagKey = "demo-feature-flag",
        treatmentVariationId = "treatment-var",
        controlVariationId = "control-var",
    )

    @Test
    fun `escapes quotes and backslashes in label values`() {
        assertEquals("""a\"b\\c""", escapePromqlLabelValue("""a"b\c"""))
    }

    @Test
    fun `builds error rate query with user context and window`() {
        val query = buildErrorRateQuery(policy, runContext, Cohort.treatment)
        assertTrue(query.contains("""ld_context_kind="user""""))
        assertTrue(query.contains("""ld_variation_id="treatment-var""""))
        assertTrue(query.contains("[300s]"))
        assertTrue(query.contains("""status=~"5..""""))
    }

    @Test
    fun `builds latency p95 query in milliseconds`() {
        val latencyPolicy = policy.copy(metricType = MetricType.latency_p95)
        val query = buildLatencyP95Query(latencyPolicy, runContext, Cohort.control)
        assertTrue(query.contains("histogram_quantile(0.95"))
        assertTrue(query.contains("""ld_variation_id="control-var""""))
        assertTrue(query.contains("* 1000"))
    }

    @Test
    fun `builds sample count query for cohorts`() {
        val query = buildSampleCountQuery(policy, runContext, Cohort.treatment)
        assertTrue(query.contains("sum(increase(http_requests_total"))
        assertTrue(query.contains("""ld_variation_id="treatment-var""""))
    }

    @Test
    fun `supports canonical metric types only`() {
        assertTrue(
            buildMetricQuery(MetricType.error_rate, policy, runContext, Cohort.treatment)
                .contains("rate(http_requests_total"),
        )
        assertTrue(
            buildMetricQuery(
                MetricType.latency_p95,
                policy.copy(metricType = MetricType.latency_p95),
                runContext,
                Cohort.treatment,
            ).contains("histogram_quantile"),
        )
        assertThrows<UnsupportedMetricTypeError> {
            buildMetricQuery("p95_latency_ms", policy, runContext, Cohort.treatment)
        }
    }
}
